apps/aecore/lib/aecore/account/account.ex

```diff
@@ -70,68 +102,32 @@ defmodule Aecore.Account.Account do
         ) :: {:ok, SignedTx.t()}
   def spend(sender, sender_priv_key, receiver, amount, fee, nonce) do
     payload = %{receiver: receiver, amount: amount}
-    spend_tx = DataTx.init(SpendTx, payload, sender, fee, nonce)
-    SignedTx.sign_tx(spend_tx, sender_priv_key)
+    spend_tx = DataTx.init(SpendTx, payload, [sender], fee, nonce)
+    SignedTx.sign_tx(spend_tx, sender, sender_priv_key)
   end
 
   @doc """
-  Adds balance to a given address (public key)
+  Adds balance to a given Account state and updates last update block.
   """
-  @spec transaction_in(Account.t(), integer()) :: Account.t()
-  def transaction_in(account_state, amount) do
+  @spec apply_transfer!(ChainState.account(), non_neg_integer(), integer()) ::
+          ChainState.account()
+  def apply_transfer!(account_state, block_height, amount) do
     new_balance = account_state.balance + amount
-    Map.put(account_state, :balance, new_balance)
-  end
 
-  @doc """
-  Deducts balance from a given address (public key)
-  """
-  @spec transaction_out(Account.t(), integer(), integer()) :: Account.t()
-  def transaction_out(account_state, amount, nonce) do
-    account_state
-    |> Map.put(:nonce, nonce)
-    |> transaction_in(amount)
-  end
-
-  @spec get_account_state(AccountStateTree.tree(), Wallet.pubkey()) :: Account.t()
-  def get_account_state(tree, key) do
-    case AccountStateTree.get(tree, key) do
-      :none ->
-        empty()
-
-      {:ok, account_state} ->
-        account_state
+    if new_balance < 0 do
+      throw({:error, "#{__MODULE__}: Negative balance"})
     end
-  end
-
-  @doc """
-  Return the balance for a given key.
-  """
-  @spec balance(AccountStateTree.tree(), Wallet.pubkey()) :: non_neg_integer()
-  def balance(tree, key) do
-    get_account_state(tree, key).balance
-  end
 
-  @doc """
-  Return the nonce for a given key.
-  """
-  @spec nonce(AccountStateTree.tree(), Wallet.pubkey()) :: non_neg_integer()
-  def nonce(tree, key) do
-    get_account_state(tree, key).nonce
+    %Account{account_state | balance: new_balance, last_updated: block_height}
   end
 
-  @doc """
-  Return the last_updated for a given key.
-  """
-  @spec last_updated(AccountStateTree.tree(), Wallet.pubkey()) :: non_neg_integer()
-  def last_updated(tree, key) do
-    get_account_state(tree, key).last_updated
-  end
+  @spec apply_nonce!(ChainState.account(), integer()) :: ChainState.account()
+  def apply_nonce!(%Account{nonce: current_nonce} = account_state, new_nonce) do
+    if current_nonce >= new_nonce do
+      throw({:error, "#{__MODULE__}: Invalid nonce"})
```

Comment:
I don't like this throw either. Why don't we just check this in `preprocess_check` or `validate` funcs? Even if there is some reason to check it again.. for whatever reason, I'm sure the `throw` could be resoved in another way.

Reply:
This is also a sanity check only. It should never throw.